### PR TITLE
DDF for iluminize Multi 5 (4A)

### PR DESCRIPTION
**Pull request by bluemoehre**
Controller by iluminize, which can emulate multiple controllers (Dimmer, CCT, RGBW, RGBCCT). 

Will solve #7985

### All Screenshots

<details><summary>Dimmer</summary>
<p>
Bildschirmfoto 2024-10-16 um 23 34 06
Bildschirmfoto 2024-10-17 um 16 02 19
Bildschirmfoto 2024-10-16 um 23 34 24
Bildschirmfoto 2024-10-16 um 23 34 38
Bildschirmfoto 2024-10-16 um 23 34 45
</p>
</details> 

<details><summary>CCT</summary>
<p>
Bildschirmfoto 2024-10-16 um 23 27 38
Bildschirmfoto 2024-10-16 um 23 27 42
Bildschirmfoto 2024-10-16 um 23 28 42
Bildschirmfoto 2024-10-16 um 23 28 51
Bildschirmfoto 2024-10-16 um 23 29 14
Bildschirmfoto 2024-10-16 um 23 29 23
Bildschirmfoto 2024-10-16 um 23 29 34
</p>
</details> 

<details><summary>RGBW</summary>
<p>
Bildschirmfoto 2024-10-16 um 22 54 23
Bildschirmfoto 2024-10-16 um 22 54 37
Bildschirmfoto 2024-10-16 um 22 54 48
Bildschirmfoto 2024-10-16 um 22 55 01
Bildschirmfoto 2024-10-16 um 22 55 12
Bildschirmfoto 2024-10-16 um 22 55 23
</p>
</details> 

<details><summary>RGBWCCT</summary>
<p>
Bildschirmfoto 2024-10-16 um 22 59 04
Bildschirmfoto 2024-10-16 um 22 59 13
Bildschirmfoto 2024-10-16 um 22 59 27
Bildschirmfoto 2024-10-16 um 22 59 36
Bildschirmfoto 2024-10-16 um 23 00 19
Bildschirmfoto 2024-10-16 um 23 00 33
Bildschirmfoto 2024-10-16 um 23 00 47
</p>
</details>

### CODE REVIEW

**Comment by bluemoehre** [comment]
> The device is missing some items because a device with "state/x" or "state/y" need the corresponding cap/color/xy/{red,green,blue}/{x,y} items

That cluster is not present. How to deal with this conflict?

**Comment by ebaauw** [comment]
Just define the capabilities with static values.

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        {
+          "name": "cap/color/xy/red_x",
+          "static": 0
+        },
+        {
+          "name": "cap/color/xy/red_y",
+          "static": 0
+        },
+        {
+          "name": "config/bri/couple_ct",
```
Comment (ebaauw) [inline devices/iluminize/led_controller_multi_5.cct.json:83]: Does the controller support this?  I’d default it to false.  Also, I would expect it to support `execute_if_off` as well.

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+          "refresh.interval": 3600,
+          "default": 336
+        },
+        {
+          "name": "config/color/execute_if_off",
+          "refresh.interval": 3600,
+          "default": true
+        },
+        {
+          "name": "config/color/xy/startup_x",
```
Comment (ebaauw) [inline devices/iluminize/led_controller_multi_5.cct.json:112]: These are only supported on Hue lights.

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+          "refresh.interval": 3600,
+          "default": "0xffff"
+        },
+        {
+          "name": "config/color/xy/startup_y",
```
Comment (ebaauw) [inline devices/iluminize/led_controller_multi_5.cct.json:117]: These are only supported on Hue lights.

File: devices/iluminize/led_controller_multi_5.dim.json [hunk 4]
```diff
@@ -0,0 +1,87 @@
+{
+  "schema": "devcap1.schema.json",
+  "manufacturername": "iluminize",
+  "modelid": "5112.80",
+  "product": "LED Controller Multi 5 (DIM mode)",
+  "sleeper": false,
+  "status": "Gold",
+  "subdevices": [
+    {
+      "type": "$TYPE_DIMMABLE_LIGHT",
```
Comment (ebaauw) [inline devices/iluminize/led_controller_multi_5.dim.json:10]: That type shouldn’t expose any colour: no `ct`, no `xy`, no `hue`/`sat` not `effect`.

File: devices/iluminize/led_controller_multi_5.rgbw.json [hunk 5]
```diff
@@ -0,0 +1,186 @@
+{
+  "schema": "devcap1.schema.json",
+  "manufacturername": "iluminize",
+  "modelid": "HK-ZD-RGB-A",
+  "product": "LED Controller Multi 5 (RGBW mode)",
```
Comment (ebaauw) [inline devices/iluminize/led_controller_multi_5.rgbw.json:5]: I don’t know the device, but typically RGBW controllers expose two endpoints: a _Color Light_ for RGB and a _Dimmable Light_ for W.  I doubt it supports `ct` in this mode.

**Review by ebaauw — COMMENTED** [review]
You didn’t post screenshots of the _Basic_ cluster, so I don’t know what ZCL version the device supports, but I would expect it to support attribute reporting.

Did you re-read the endpoints and _Simple Descriptor_ each time after changing the mode?  It might not make sense to expose an RGBW light as _Extended Color Light_.  The colour temperatures will be ugly, mixing RGB and W.

File: devices/iluminize/led_controller_multi_5.rgbw.json (hunk 5 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.rgbw.json:5]: These endpoints are active and show changing numbers. If I remember correctly they are not writable. The controller seems to do some internal mix of colors and is optimized for neutral white LEDs (4000 kelvin). All Philips RGBW lamps I had up to now do the same and don't provide a dedicated white dimmer.

File: devices/iluminize/led_controller_multi_5.cct.json (hunk 1 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:83]: If you could tell me the cluster + attribute address I will re-check. 

BTW: I find it very hard with the DDF editor to match against the controller. My workflow is always drop everything in, then compare endpoints if active. It takes too long and easily causes to miss something. Is there a better way to match the written path vs available clusters/addresses?

**Comment by bluemoehre** [comment]
@ebaauw Seems like my idea to copy from another DDF was not the best. Since we have the bot checking DDF files. Are you able to add these hints/rules to it? Then we can run it vs all existing files and do a clean up? This way the next person won't trap into the same problems.

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+          "refresh.interval": 3600,
+          "default": "0xffff"
+        },
+        {
+          "name": "config/colorcapabilities",
+          "refresh.interval": 86400
+        },
+        {
+          "name": "config/ctmax",
```
[the remark this answers is not in the document]
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:126]: I guess I added that one in the DDF editor. So it's still available there? Are deprecated attributes marked in some way, so I can re-check?

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+          "refresh.interval": 3600,
+          "default": "0xffff"
+        },
+        {
+          "name": "config/color/xy/startup_y",
+          "refresh.interval": 3600,
+          "default": "0xffff"
+        },
+        {
+          "name": "config/colorcapabilities",
```
[the remark this answers is not in the document]
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:122]: (… same as before)

File: devices/iluminize/led_controller_multi_5.rgbw.json (hunk 5 above)
Comment (ebaauw) [reply devices/iluminize/led_controller_multi_5.rgbw.json:5]: Correct, the endpoints and clusters are reported by the device and cannot be written.

What Philips lights are you referring to?  Afaik, Hue don’t have RGBW lights.

There are other LED controllers that use multiple endpoints, like GLEDOPTO.


But anyways, if the device supports `ct` in  RGBW mode, we should expose it as a _Color Dimmable Light_.

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+          "refresh.interval": 86400
+        },
+        {
+          "name": "config/ctmax",
+          "default": "0xfeff"
+        },
+        {
+          "name": "config/ctmin",
```
[the remark this answers is not in the document]
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:130]: (… same as before)

File: devices/iluminize/led_controller_multi_5.cct.json (hunk 2 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:112]: (… same as below)

**Comment by bluemoehre** [comment]
> You didn’t post screenshots of the _Basic_ cluster, so I don’t know what ZCL version the device supports, but I would expect it to support attribute reporting.

I already mounted it behind a wall ... but I gonna try to upload these as well.

> Did you re-read the endpoints and _Simple Descriptor_ each time after changing the mode? It might not make sense to expose an RGBW light as _Extended Color Light_. The colour temperatures will be ugly, mixing RGB and W.

Since re-reading didn't work all time, causing deCONZ to show old stuff, I deleted the node from the network and then hard reset the controller each time.

File: devices/iluminize/led_controller_multi_5.cct.json (hunk 3 above)
Comment (ebaauw) [reply devices/iluminize/led_controller_multi_5.cct.json:117]: Added what?  HomeKit doesn’t have poweron settings, and certainly not for colour.  I’ve seen some vendors that (illegally) expose the Hue _Manufacturer Code_ 0x100B in the _Node Descriptor_ (also missing from the screenshots), so the Hue bridge exposes them to HomeKit.  In that case, they might expose the Hue-specific attributes for poweron xy as well.  However, they are not in the screenshots?  Do you see them in the GUI?  And do they work?

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 9]
```diff
@@ -55,6 +55,30 @@
...
+          "name": "cap/color/xy/blue_y",
+          "static": 0
+        },
+        {
+          "name": "cap/color/xy/green_x",
+          "static": 0
+        },
+        {
+          "name": "cap/color/xy/green_y",
+          "static": 0
```
[the remark this answers is not in the document]
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:72]: I see ... I think I copied that from somewhere. I will correct this and anything similar.

File: devices/iluminize/led_controller_multi_5.rgbw.json (hunk 5 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.rgbw.json:5]: > What Philips lights are you referring to? Afaik, Hue don’t have RGBW lights.

In some Philips HUE I disassembled also have been RGBW LEDs (maybe it was Xamento), but the controller does not provide a dedicated cluster to control the white ones. So I assume it also does some mix which matches the colorspace of the LEDs. (Annoying here: in white colors it's freaking bright and while using saturated colors, it is very dark.)

**Comment by bluemoehre** [comment]
added Basic cluster for Dimmer Mode.

**Comment by ebaauw** [comment]
@Zehir maintains the DDF validator; I haven’t looked into its internals. It does check for mandatory items, but I don’t think it currently checks for superfluous or deprecated items.  Of course, the validator cannot check against the actual device features.

I often copy an existing DDF as a start to create new DDF, but I do check the validity and appropriateness of each item in the GUI.  There’s also some hidden (from API clients) capabilities that change the API behaviour, like using _Move to Level (with On/Off)_ or _Off with Effect_.  You need to check in the GUI if the device supports these.  There’s a brief description in each item’s JSON file (under `devices/generic/items/*_item.json`) but there’s definitely room to improve the documentation.  The use of the red, green, and blue points, to specify what colors the light can display, is best explained in the Hue API documentation.

**Comment by ebaauw** [comment]
> added Basic cluster for Dimmer Mode (2nd image in PR description 

ZCL Version 3 is Zigbee 3 and should support attribute reporting.

**Comment by bluemoehre** [comment]
> ZCL Version 3 is Zigbee 3 and should support attribute reporting.

So we need to additionally add some reporting config everywhere?

File: devices/iluminize/led_controller_multi_5.dim.json (hunk 4 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.dim.json:10]: You're right. Gonna fix that one (and wonder how candle / fire effects in non-color controllers would look like)

File: devices/iluminize/led_controller_multi_5.cct.json [hunk 10]
```diff
@@ -0,0 +1,186 @@
+{
+  "schema": "devcap1.schema.json",
+  "manufacturername": "iluminize",
+  "modelid": "HK-ZD-CCT-A",
+  "product": "LED Controller Multi 5 (CCT mode)",
+  "sleeper": false,
+  "status": "Gold",
+  "subdevices": [
+    {
+      "type": "$TYPE_COLOR_TEMPERATURE_LIGHT",
```
[the remark this answers is not in the document]
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:10]: I checked this again, but the controller does provide these as shown in the screenshot. 

In terms of colorspaces it is absolutely possible to convert CCT into HUE and XY. So maybe it is just the first we see doing it?

For the effects I wonder anyhow why there shouldn't be some on dimmable & CCT controllers. Candle / fire flicker would be possible =)

File: devices/iluminize/led_controller_multi_5.cct.json (hunk 1 above)
Comment (bluemoehre) [reply devices/iluminize/led_controller_multi_5.cct.json:83]: I double checked, but it is as displayed in the screenshots:

Level Control Options bitmap is: `0x01`
Color Control Options bitmap is: `0x01`

do I need to adjust something?

**Review by SwoopX — CHANGES_REQUESTED** [review]
Hi and thank you for your contribution 👍 

We would like to ask you to create a new device support request (new issue -> device support) in this repository and to link it against your pull request. Please also make sure one device support PR only covers one device. This fosters consistency, allowing everybody to find device related information at the very same place and helps expediting searches if required.

Thank you very much for your support!

**Comment by bluemoehre** [comment]
> We would like to ask you to create a new device support request (new issue -> device support) in this repository and to link it against your pull request. 

done

**Comment by SwoopX** [comment]
Thanks a lot 👍

**Review by SwoopX — APPROVED** [review]

**Comment by manup** [comment]
@ebaauw is the PR ready for merging?

**Comment by ebaauw** [comment]
I suppose so.  I don't have the device so I cannot validate the DDF.

**Comment by bluemoehre** [comment]
@ebaauw it would help (me) to understand the status of an issue, if you could resolve all started conversations, where you think it now looks good. This prevents to miss something we had been discussing without having a solution. (Additionally I am not sure if someone could merge with unsolved discussions.)

I will also check myself everything here and in other tickets …

**Comment by ebaauw** [comment]
Since the DDF (by design) won't affect any other devices, merging an incomplete or even wrong DDF would cause no harm, other than the device not working as expected.

The idea of a DDF, particularly of the `capabilities` items, is to indicate what actually works for this device, not just what attributes it advertises.  Checking this requires trying out the features in the GUI, and visually checking the light's reaction.  Support for attribute reporting can be checked by the GUI showing the changed values automatically (without having to re-read the attributes), through the deCONZ log, or using a Zigbee sniffer.

So for example, if the DDF advertises the device supports _Execute If Off_, the deCONZ REST API assumes it does, and will happily send a _Move to Level_ while the device is off, and update `state/bri` accordingly (rather than raising an API error that the light is off).  If the light subsequently is turned on, but doesn't reflect the changed brightness it doesn't live up to the promise the DDF makes to API clients.

**Pull request merged**